src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java

```diff
@@ -19,26 +27,69 @@
  */
 public class AddRecordsProcessor extends SinkRecordsProcessor {
 
-	public AddRecordsProcessor(Queue<Collection<SinkRecord>> sinkRecords, AtomicBoolean shouldRun,
-			Repository repository, RDFFormat format, int batchSize, long timeoutCommitMs) {
-		super(sinkRecords, shouldRun, repository, format, batchSize, timeoutCommitMs);
+	public AddRecordsProcessor(Queue<Collection<SinkRecord>> sinkRecords, AtomicBoolean shouldRun, Repository repository,
+							   RDFFormat format, int batchSize, long timeoutCommitMs, ErrorHandler errorHandler) {
+		super(sinkRecords, shouldRun, repository, format, batchSize, timeoutCommitMs, errorHandler);
 	}
 
 	@Override
 	public void flushRecordUpdates() {
-		//no need to create connection if already empty
-		if (!recordsBatch.isEmpty()) {
+		try {
+			if (!recordsBatch.isEmpty()) {
+				flush();
+				recordsBatch.clear();
+			}
+		} catch (TimeoutException e) {
+			LogErrorHandler.logError(e.getMessage());
+			throw new RuntimeException(e);
+		}
+	}
+
+	private void flush() throws TimeoutException {
+		long retryCount = 1L;
+		Queue<SinkRecord> failedRecords = new ArrayDeque<>(recordsBatch.size());
+
+		do {
+			if (!failedRecords.isEmpty()) {
+				removeFailed(failedRecords);
```

Comment:
And if the batch size is 1 000 000, you would have to do linear searches every time. Remove failed records queue and handle each record directly by simply removing it and processing it with the error handler.

Reply:
IF you have 100 failed records and
IF you get an IOException at 999 999 record 
    than you will remove all failed 100 records from the Queue and retry the connection as long as it takes with all 999 **8**99 valid records.